### PR TITLE
refactor(robot-server): ensure persisted datetimes are extracted as UTC

### PR DESCRIPTION
**Pull request by mcous**
## Overview

SQLAlchemy when used with SQLite is unable to preserve timezone information when inserting datetimes into the database. Instead, they are stored as "naive" (timezone-less) datetimes.

Since I couldn't find any way to configure timezones to be stored, this PR adds on UTC to any naive datetime extracted from the DB.

## Changelog

- refactor(robot-server): ensure persisted datetimes are extracted as UTC

## Review requests

- [ ] Observed client bugs due to timezone problems are no longer observed

## Risk assessment

Low. We are very good about ensuring we always attach UTC to created datetimes, _and_ the robot has its timezone set to UTC, so even if any locale-based datetimes sneak in, they'd still be UTC in production.

### CODE REVIEW

**Review by b-cooper — APPROVED** [review]
Change looks good. Tested on 🤖 and the run log behaves as expected for calculating command and run durations.

File: robot-server/robot_server/persistence.py [hunk 1]
```diff
@@ -36,7 +35,8 @@
     ),
     sqlalchemy.Column(
         "created_at",
-        sqlalchemy.DateTime,
+        # NOTE: SQLite does not support timezone-aware datetimes
+        sqlalchemy.DateTime(timezone=True),
```
Comment (SyntaxColoring) [inline robot-server/robot_server/persistence.py]: I'm not sure I understand what this comment is saying? And it looks like this `timezone=True` is a no-op—is that intentional?

My thinking:

In a sense, SQLite doesn't "support" datetimes *at all.* It's just strings. From [Dialects > SQLite > Date and Time Types](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#date-and-time-types) (emphasis mine):

> SQLite does not have built-in `DATE`, `TIME`, or `DATETIME` types.... SQLAlchemy’s own [DateTime](https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.DateTime) and related types provide date formatting and parsing functionality when SQLite is used. The implementation classes are [DATETIME](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME), [DATE](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATE) and [TIME](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlalchemy.dialects.sqlite.TIME). **These types represent dates and times as ISO formatted strings**

And, scrolling down to [sqlalchemy.dialects.sqlite.DATETIME](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME):

> Represent a Python datetime object in SQLite using a string.
>
> The default string storage format is:
>
> ```
> "%(year)04d-%(month)02d-%(day)02d %(hour)02d:%(minute)02d:%(second)02d.%(microsecond)06d"
> ```
>
> [note no timezone]

My reading of all of this is that SQLAlchemy maps `datetime`s to `str`s and stores the `str`s in SQLite. But the `str` format that SQLAlchemy uses has no place for a timezone. So it's not exactly that SQLite doesn't support timezone-aware `datetime`s—it's that SQLAlchemy's back-end for SQLite drops timezones when it stores `datetime`s in the database.

With this in mind, it also seems like `timezone=True` is ultimately a no-op here. From [its documentation](https://docs.sqlalchemy.org/en/14/core/type_basics.html?highlight=datetime#sqlalchemy.types.DateTime.params.timezone) (emphasis theirs):

> timezone – boolean. Indicates that the datetime type should enable timezone support, if available on the **base date/time-holding type only.**

If the "base date/time-holding type" is [`sqlalchemy.dialects.sqlite.DATETIME`](https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME), I don't think timezone support *is* available on it. This is apparently [quite an old SQLAlchemy issue](https://github.com/sqlalchemy/sqlalchemy/issues/1985).

**Review by SyntaxColoring — COMMENTED** [review]
I think I'm 50% following.

I see how this fixes up `datetime`s that we read from the database, but what about `datetime`s that we write to the database?

File: robot-server/robot_server/persistence.py (hunk 1 above)
Comment (TamarZanzouri) [reply robot-server/robot_server/persistence.py]: Should we change this to a str and handle conversion on insert and get?

Comment (mcous) [reply robot-server/robot_server/persistence.py]: My intention with the (non-functional) option and the note comment was to try to say "we intend for the inserted datetimes to have timezone info, _and_ we know SQLite won't do that". To a much lesser extent I also wanted to cover our bases for a DB backend swap. Open to change anything here.

Weirdly (to me, after reading the docs), if you open the SQLite in a DB browser, the column type does show up as "DATETIME", so I'm a little mystified in general here.

**Comment by mcous** [comment]
> I think I'm 50% following.
> 
> 
> 
> I see how this fixes up `datetime`s that we read from the database, but what about `datetime`s that we write to the database?

This PR works because we always create datetimes with the UTC timezones (by convention) before writing. When a datetime gets written to the DB the tzinfo is simply stripped, which makes it relatively safe to tack UTC onto the naive datetime on extraction. It's a relatively fragile chain at the moment, but I think it can be vastly improved by more ORM functionality.

**Review by mcous — COMMENTED** [review]
TODO from review meeting

- [x] Remove useless `timezone=True`
- [x] Expand comment to note that input datetimes _must_ be UTC
    - Clarify what "timezones are not supported" actually means
- [x] Explore anything lightweight for this PR that could assert/validate that input timestamps are UTC
    - Explore adding it to "input" classes

**Review by SyntaxColoring — APPROVED** [review]
I haven't tested this because it's unclear to me how to reproduce the original bug, but these code changes look good to me. Thanks!

**Review by TamarZanzouri — APPROVED** [review]

**Pull request merged**